v2/datastream-to-spanner/terraform/samples/all-infrastructure/README.md

```diff
@@ -0,0 +1,554 @@
+   to the MySQL instance (using IP whitelisting).
+3. **GCS bucket** - A GCS bucket to for Datastream to write the source data to.
+4. **Target datastream connection profile** - The connection profile to
+   configure the created bucket in Datastream.
+5. **Pubsub topic and subscription** - This contains GCS object notifications as
+   files are written to GCS for consumption by the Dataflow job.
+6. **Datastream stream** - A datastream stream which reads from the source
+   specified in the source connection profile and writes the data to the bucket
+   specified in the target connection profile. Note that it uses a mandatory
+   prefix path inside the bucket where it will write the data to. The default
+   prefix path is `data` (can be overridden).
+7. **Bucket notification** - Creates the GCS bucket notification which publish
+   to the pubsub topic created. Note that the bucket notification is created on
+   the mandatory prefix path specified for the stream above.
+8. **Dataflow job** - The Dataflow job which reads from GCS and writes to
+   Spanner.
+9. **Permissions** - It adds the required roles to the specified (or the
+   default) service accounts for the live migration to work.
+
+> **_NOTE:_** A label is attached to all the resources created via Terraform.
+> The key is `migration_id` and the value is auto-generated. The auto-generated
+> value is used as a global identifier for a migration job across resources. The
+> auto-generated value is always pre-fixed with a `smt-`.
+
+## How to run
+
+1. Clone this repository or the sample locally.
+2. Edit the `terraform.tfvars` or `terraform_simple.tfvars` file and replace the
+   dummy variables with real values. Extend the configuration to meet your
+   needs. It is recommended to get started with `terraform_simple.tfvars`.
+3. Run the following commands -
+
+### Initialise Terraform
+
+```shell
+# Initialise terraform - You only need to do this once for a directory.
+terraform init
+```
+
+### Run `plan` and `apply`
+
+Validate the terraform files with -
+
+```shell
+terraform plan --var-file=terraform_simple.tfvars
+```
+
+Run the terraform script with -
+
+```shell
+terraform apply --var-file=terraform_simple.tfvars
+```
+
+This will launch the configured jobs and produce an output like below -
+
+```shell
+Outputs:
+
+resource_ids = {
+  "dataflow_job" = "2024-06-14_03_01_00-3421054840094926119"
+  "datastream_source_connection_profile" = "source-mysql-thorough-wombat"
+  "datastream_stream" = "mysql-stream-thorough-wombat"
+  "datastream_target_connection_profile" = "target-gcs-thorough-wombat"
+  "gcs_bucket" = "live-migration-thorough-wombat"
+  "pubsub_subscription" = "live-migration-thorough-wombat-sub"
+  "pubsub_topic" = "live-migration-thorough-wombat"
+}
+resource_urls = {
+  "dataflow_job" = "https://console.cloud.google.com/dataflow/jobs/us-central1/2024-06-14_03_01_00-3421054840094926119?project=your-project-here"
```

Comment:
nit: let's replace the params with a clearer seperator. Something like <your-project-here> or so ?

Reply:
Also, the dataflow job url's will depend on the region where this job is being run.

Reply:
> Also, the dataflow job url's will depend on the region where this job is being run.

No, it doesn't.

> nit: let's replace the params with a clearer seperator. Something like or so ?

These are clickable hyperlinks and are pretty intuitive to understand from the output. Keeping them as is.